tests/components/sensor/test_foobot.py

```diff
@@ -0,0 +1,75 @@
+"""The tests for the Foobot sensor platform."""
+
+import re
+import unittest
+
+import homeassistant.components.sensor as sensor
+from homeassistant.const import (TEMP_CELSIUS)
+from homeassistant.exceptions import PlatformNotReady
+from homeassistant.setup import setup_component
+from tests.common import (
+    assert_setup_component, get_test_home_assistant, load_fixture)
+
+from ...test_util.aiohttp import mock_aiohttp_client
+
+VALID_CONFIG = {
+    'platform': 'foobot',
+    'token': 'adfdsfasd',
+    'username': 'example@example.com',
+}
+
+
+# pylint: disable=invalid-name
+class TestFoobotSetup(unittest.TestCase):
```

Comment:
Since all your code is async, prefer to write async test code too. That will also allow you to leverage the `hass` and `aioclient_mock` fixtures. [Example](https://github.com/home-assistant/home-assistant/blob/dev/tests/helpers/test_aiohttp_client.py#L151-L155)

Replace setup_component with async_setup_component and use `assert X == Y` instead of `assertEquals`

```python
async def test_default_setup(hass, aioclient_mock):
    aioclient_mock.get(…)
    assert await async_setup_component(…)
```

Reply:
Thanks for this example, the test is much cleaner and simpler now.

But unfortunately, I've run into an issue while trying to test for errors:

```python
async def test_setup_error(hass, aioclient_mock):
    """Expected failures caused by various errors in API response."""
    aioclient_mock.get(re.compile('api.foobot.io/v2/owner/.*'),
                       exc=asyncio.TimeoutError())

    with pytest.raises(PlatformNotReady):
        await async_setup_component(hass, sensor.DOMAIN,
                                    {'sensor': VALID_CONFIG})
```

Running the test returns:

```
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_setup_error[pyloop] ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

hass = <homeassistant.core.HomeAssistant object at 0x10f0edc88>, aioclient_mock = <tests.test_util.aiohttp.AiohttpClientMocker object at 0x10f242dd8>

    async def test_setup_error(hass, aioclient_mock):
        """Expected failures caused by various errors in API response."""
        aioclient_mock.get(re.compile('api.foobot.io/v2/owner/.*'),
                           exc=asyncio.TimeoutError())
    
        with pytest.raises(PlatformNotReady):
            await async_setup_component(hass, sensor.DOMAIN,
>                                       {'sensor': VALID_CONFIG})
E           Failed: DID NOT RAISE <class 'homeassistant.exceptions.PlatformNotReady'>

tests/components/sensor/test_foobot.py:52: Failed
----------------------------------------------------------------- Captured stderr call ------------------------------------------------------------------
ERROR:homeassistant.components.sensor.foobot:Failed to connect to foobot servers.
Traceback (most recent call last):
  File "/Users/reefab/Projects/Public/home-assistant/homeassistant/components/sensor/foobot.py", line 67, in async_setup_platform
    devices = await client.get_devices()
  File "/Users/reefab/Projects/Public/home-assistant/lib/python3.6/site-packages/foobot_async/__init__.py", line 57, in get_devices
    return (yield from self._get(DEVICE_URL.format(username= self._username)))
  File "/Users/reefab/Projects/Public/home-assistant/lib/python3.6/site-packages/foobot_async/__init__.py", line 153, in _get
    path, headers=dict(self._headers, **kwargs))
  File "/Users/reefab/Projects/Public/home-assistant/tests/test_util/aiohttp.py", line 102, in match_request
    raise response.exc
concurrent.futures._base.TimeoutError
WARNING:homeassistant.components.sensor:Platform foobot not ready yet. Retrying in 30 seconds.
------------------------------------------------------------------- Captured log call -------------------------------------------------------------------
foobot.py                   79 ERROR    Failed to connect to foobot servers.
Traceback (most recent call last):
  File "/Users/reefab/Projects/Public/home-assistant/homeassistant/components/sensor/foobot.py", line 67, in async_setup_platform
    devices = await client.get_devices()
  File "/Users/reefab/Projects/Public/home-assistant/lib/python3.6/site-packages/foobot_async/__init__.py", line 57, in get_devices
    return (yield from self._get(DEVICE_URL.format(username= self._username)))
  File "/Users/reefab/Projects/Public/home-assistant/lib/python3.6/site-packages/foobot_async/__init__.py", line 153, in _get
    path, headers=dict(self._headers, **kwargs))
  File "/Users/reefab/Projects/Public/home-assistant/tests/test_util/aiohttp.py", line 102, in match_request
    raise response.exc
concurrent.futures._base.TimeoutError
entity_platform.py          99 WARNING  Platform foobot not ready yet. Retrying in 30 seconds.

 tests/components/sensor/test_foobot.py ⨯                                                                                                 100% ██████████

Results (0.44s):
       1 passed
       1 failed
         - tests/components/sensor/test_foobot.py:45 test_setup_error[pyloop]
sys:1: RuntimeWarning: coroutine 'EntityPlatform.async_setup' was never awaited
```

For some reason, having an `await` inside a `with purest.raises` seems to mess things up and I always end up with a ` RuntimeWarning: coroutine 'EntityPlatform.async_setup' was never awaited`.

I'll revert this particular test to the old synchronous way for now.

Reply:
A call to `async_setup_component` will never raise `PlatformNotReady`. That exception will be handled internally.

Reply:
So you will have to call `async_setup_platform` directly in that case.

Reply:
Thanks, that was the piece of the puzzle I was missing. I did indeed get confused between `async_setup_component` and `async_setup_platform`.

 I fixed that and added tests for failure scenarios.